lib/bootsnap/load_path_cache/cache.rb

```diff
@@ -9,7 +9,7 @@ def initialize(store, path_obj, development_mode: false)
         @development_mode = development_mode
         @store = store
         @mutex = defined?(::Mutex) ? ::Mutex.new : ::Thread::Mutex.new # TODO: Remove once Ruby 2.2 support is dropped.
-        @path_obj = path_obj
+        @path_obj = path_obj.map { |f| File.exist?(f) ? File.realpath(f) : f }
```

Comment:
Ah, this is the problem! `@path_obj` needs to continue to be an actual reference to `$LOAD_PATH` or to `ActiveSupport::Dependencies.autoload_paths`, and any change to it needs to trigger a corresponding event in `ChangeObserver`.

One way to do this here would be to:

```ruby
        path_obj.map! { |f| File.exist?(f) ? File.realpath(f) : f }
        @path_obj = path_obj
```

Reply:
It would be best too, at this point, to either:

1. Make the `ChangeObserver` understand `map!`; or
2. Make this line blow up if the `ChangeObserver` has bound to this object`.

Reply:
ouch, understood, will take a look.

Reply:
AFAICS ChangeObserver [already understands `map!`](https://github.com/Shopify/bootsnap/blob/master/lib/bootsnap/load_path_cache/change_observer.rb#L44), so `.map!` and we should be fine?
Any hints how it can be reproduced locally?

Reply:
Oh great then.

To reproduce, I guess:

1. load bootsnap
2. Add a new dir to the load path via, say, `$LOAD_PATH.unshift`
3. Observe that bootsnap will not load features from this path since the `ChangeObserver`  hook did not fire.

An easy way to catch this is to just log or whatever from the hooked method in `ChangeObserver`.